From e328da1be2186ccc87f196a7156e23b3d1b71cb9 Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Mon, 30 Jan 2012 20:57:55 +0100
Subject: [PATCH 021/124] MediaMonitor doesn't handle 'Jump to plugin'
 correctly

In commit ed4d961aed836fade236d006b2797f0c044970de the call to
GetMythMainWindow()->JumpTo("Main Menu") was moved from
MediaMonitor::JumpToMediaHandler into the plugins MythMusic, MythVideo &
MythGallery.

However, JumpTo() doesn't complete synchronously, it calls
QCoreApplication::postEvent to do the work.  When the plugin starts its
event loop it finds the jump event and immediately exits without playing
the inserted media.

It is necessary to execute the following code after calling JumpTo():

     GetMythMainWindow()->JumpTo("Main Menu");
     QTime t; t.start();
     while (GetMythMainWindow()->IsExitingToMain() && t.elapsed() < 2000)
         qApp->processEvents(); // Ensure jump is executed before calling handler

Because of this it is preferable to keep the call to JumpTo() in
MediaMonitor::JumpToMediaHandler and append the above code rather than
replicate it in all plugins.

Ticket URL: <http://code.mythtv.org/trac/ticket/9537>

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythplugins/mythgallery/mythgallery/main.cpp |    1 -
 mythplugins/mythmusic/mythmusic/main.cpp     |    2 --
 mythtv/libs/libmyth/mythmediamonitor.cpp     |   11 ++++++++++-
 mythtv/programs/mythfrontend/main.cpp        |    2 --
 4 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/mythplugins/mythgallery/mythgallery/main.cpp b/mythplugins/mythgallery/mythgallery/main.cpp
index 396c9d6..c6d39d5 100644
--- a/mythplugins/mythgallery/mythgallery/main.cpp
+++ b/mythplugins/mythgallery/mythgallery/main.cpp
@@ -77,7 +77,6 @@ static void handleMedia(MythMediaDevice *dev)
 
     if (dev && dev->isUsable())
     {
-        GetMythMainWindow()->JumpTo("Main Menu");
         run(dev);
     }
 }
diff --git a/mythplugins/mythmusic/mythmusic/main.cpp b/mythplugins/mythmusic/mythmusic/main.cpp
index 7c5ef9d..1701a5b 100644
--- a/mythplugins/mythmusic/mythmusic/main.cpp
+++ b/mythplugins/mythmusic/mythmusic/main.cpp
@@ -522,8 +522,6 @@ static void handleMedia(MythMediaDevice *cd)
         return;
     }
 
-    GetMythMainWindow()->JumpTo("Main Menu");
-
     if (gCoreContext->GetNumSetting("AutoPlayCD", 0))
     {
         // Empty the playlist to ensure CD is played first
diff --git a/mythtv/libs/libmyth/mythmediamonitor.cpp b/mythtv/libs/libmyth/mythmediamonitor.cpp
index 853095d..1666c24 100644
--- a/mythtv/libs/libmyth/mythmediamonitor.cpp
+++ b/mythtv/libs/libmyth/mythmediamonitor.cpp
@@ -370,6 +370,9 @@ bool MediaMonitor::RemoveDevice(const QString &dev)
     {
         if ((*it)->getDevicePath() == dev)
         {
+            // Ensure device gets an unmount
+            (*it)->checkMedia();
+
             if (m_UseCount[*it] == 0)
             {
                 (*it)->deleteLater();
@@ -674,6 +677,10 @@ void MediaMonitor::JumpToMediaHandler(MythMediaDevice* pMedia)
     // if user didn't cancel, selected = handlers.at(choice);
     int selected = 0;
 
+    GetMythMainWindow()->JumpTo("Main Menu");
+    QTime t; t.start();
+    while (GetMythMainWindow()->IsExitingToMain() && t.elapsed() < 2000)
+        qApp->processEvents(); // Ensure jump is executed before calling handler
     handlers.at(selected).callback(pMedia);
 }
 
@@ -697,7 +704,9 @@ void MediaMonitor::mediaStatusChanged(MythMediaStatus oldStatus,
     // This gets called from outside the main thread so we need
     // to post an event back to the main thread.
     // We now send events for all non-error statuses, so plugins get ejects
-    if (stat != MEDIASTAT_ERROR && stat != MEDIASTAT_UNKNOWN)
+    if (stat != MEDIASTAT_ERROR && stat != MEDIASTAT_UNKNOWN &&
+        // Don't send an event for a new device that's not mounted
+        !(oldStatus == MEDIASTAT_UNPLUGGED && stat == MEDIASTAT_NOTMOUNTED))
     {
         // Should we ValidateAndLock() first?
         QEvent *e = new MythMediaEvent(stat, pMedia);
diff --git a/mythtv/programs/mythfrontend/main.cpp b/mythtv/programs/mythfrontend/main.cpp
index 244db78..09c6f20 100644
--- a/mythtv/programs/mythfrontend/main.cpp
+++ b/mythtv/programs/mythfrontend/main.cpp
@@ -789,10 +789,8 @@ static void handleDVDMedia(MythMediaDevice *dvd)
         case 0 : // Do nothing
             break;
         case 1 : // Display menu (mythdvd)*/
-            GetMythMainWindow()->JumpTo("Main Menu");
             break;
         case 2 : // play DVD or Blu-ray
-            GetMythMainWindow()->JumpTo("Main Menu");
             playDisc();
             break;
         default:
-- 
1.7.9.5

